pkg/cloudprovider/aws/pricing.go

```diff
@@ -0,0 +1,363 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aws
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"go.uber.org/multierr"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/ec2/ec2iface"
+	"github.com/aws/aws-sdk-go/service/pricing"
+	"github.com/aws/aws-sdk-go/service/pricing/pricingiface"
+	"github.com/samber/lo"
+	"knative.dev/pkg/logging"
+)
+
+// PricingProvider provides actual pricing data to the AWS cloud provider to allow it to make more informed decisions
+// regarding which instances to launch.  This is initialized at startup with a periodically updated static price list to
+// support running in locations where pricing data is unavailable.  In those cases the static pricing data provides a
+// relative ordering that is still more accurate than our previous pricing model.  In the event that a pricing update
+// fails, the previous pricing information is retained and used which may be the static initial pricing data if pricing
+// updates never succeed.
+type PricingProvider struct {
+	ec2     ec2iface.EC2API
+	pricing pricingiface.PricingAPI
+	region  string
+
+	mu                 sync.RWMutex
+	onDemandUpdateTime time.Time
+	onDemandPrices     map[string]float64
+	spotUpdateTime     time.Time
+	spotPrices         map[string]float64
+}
+
+// pricingUpdatePeriod is how often we try to update our pricing information after the initial update on startup
+const pricingUpdatePeriod = 12 * time.Hour
+
+// NewPricingAPI returns a pricing API configured based on a particular region
+func NewPricingAPI(sess *session.Session, region string) pricingiface.PricingAPI {
+	if sess == nil {
+		return nil
+	}
+	// pricing API doesn't have an endpoint in all regions
+	pricingAPIRegion := "us-east-1"
+	if strings.HasPrefix(region, "cn-") {
+		pricingAPIRegion = "cn-north-1"
+	} else if strings.HasPrefix(region, "ap-") {
+		pricingAPIRegion = "ap-south-1"
+	}
```

Comment:
Not sure I understand this - there's 6 regions that start with `ap` so wondering why we're redirecting all their traffic to `ap-south-1` ? Is this just for lower latencies or something?

Reply:
The only two endpoints listed at https://docs.aws.amazon.com/general/latest/gr/billing.html for pricing are us-east-1 and ap-south-1.